Fix compatibility with starlette 0.18.0

TimPansino opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

*

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

patrick91 commented on strawberry/asgi/handlers/http_handler.py:
```diff
@@ -54,7 +48,12 @@ async def handle(self, scope: Scope, receive: Receive, send: Send):
         )
 
-        response.headers.raw.extend(sub_response.headers.raw)
+        # Only add headers from sub_response if they're not already present in response
+        response.headers.raw.extend(
+            (k, v)
+            for k, v in sub_response.headers.raw
+            if k not in {k for k, _ in response.headers.raw}
+        )
```
should we add a warning if we are trying to override some headers?

TimPansino replied:
Up to you, this is currently blocking `content-length: 0` from appearing and being a duplicate which also crashes starlette.

patrick91 replied:
let's do that if you don't mind 😊 I think it might be helpful in cases where you might set headers twice

TimPansino commented:
All tests passed locally, looks like some tests failed on the runner though.

```
FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_missing_query - a...
FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_not_sending_wrong_content_type
FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_method_is_not_allowed[PUT]
FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_method_is_not_allowed[DELETE]
FAILED 😰  tests/asgi/test_query.py::test_fails_when_request_body_has_invalid_json
FAILED 😰  tests/asgi/test_view.py::test_renders_graphiql_disabled[/] - asser...
FAILED 😰  tests/asgi/test_view.py::test_renders_graphiql_disabled[/graphql]
```

patrick91 commented:
> All tests passed locally, looks like some tests failed on the runner though.
> 
> ```
> FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_missing_query - a...
> FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_not_sending_wrong_content_type
> FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_method_is_not_allowed[PUT]
> FAILED 😰  tests/asgi/test_http.py::test_returns_error_when_method_is_not_allowed[DELETE]
> FAILED 😰  tests/asgi/test_query.py::test_fails_when_request_body_has_invalid_json
> FAILED 😰  tests/asgi/test_view.py::test_renders_graphiql_disabled[/] - asser...
> FAILED 😰  tests/asgi/test_view.py::test_renders_graphiql_disabled[/graphql]
> ```

Could be due to the different version of starlette on CI?

TimPansino commented:
@patrick91 

* I've fixed those failing tests on starlette versions 0.16, 0.17, and 0.18. The only tests that fail when I run locally are FastAPI for starlette 0.18, which those two are incompatible so that's not really a strawberry problem.
* I updated pyproject.toml to reflect the new versions of starlette that should be supported.
* I added the warning message for duplicate headers, needed an actual loop which is almost certainly slower but may be a better user experience. 

Let me know if there are any other changes you want made to this PR.

If I may suggest something, your testing could be made much more thorough if you used a tool like `tox` or `nox` to test each of your compatibility layers with a matrix of package and python versions. Obviously that bloats testing times on actions but is much more thorough. If you have any interest feel free to check out https://github.com/newrelic/newrelic-python-agent for an example of how we use this for testing.

patrick91 commented:
> @patrick91
> 
>     * I've fixed those failing tests on starlette versions 0.16, 0.17, and 0.18. The only tests that fail when I run locally are FastAPI for starlette 0.18, which those two are incompatible so that's not really a strawberry problem.
> 
>     * I updated pyproject.toml to reflect the new versions of starlette that should be supported.

Awesome! 

>     * I added the warning message for duplicate headers, needed an actual loop which is almost certainly slower but may be a better user experience.

Yes! Hopefully people won't have too many headers :D

 
> Let me know if there are any other changes you want made to this PR.

I'll check it soon!

> If I may suggest something, your testing could be made much more thorough if you used a tool like `tox` or `nox` to test each of your compatibility layers with a matrix of package and python versions. Obviously that bloats testing times on actions but is much more thorough. If you have any interest feel free to check out https://github.com/newrelic/newrelic-python-agent for an example of how we use this for testing.

We do something similar to this for django, but maybe tox is easier to use. Maybe we could mark the tests with `pytest.mark.starlette` and so on so we only test relevant tests when using the matrix?

patrick91 commented:
I wonder how we can circumvent this issue with FastAPI not supporting the latest version of starlette 🤔

patrick91 commented:
I guess we can wait for this PR to be merged and released 😊 https://github.com/tiangolo/fastapi/pull/4483

Kludex commented on strawberry/asgi/handlers/http_handler.py:
```diff
@@ -35,13 +36,10 @@ async def handle(self, scope: Scope, receive: Receive, send: Send):
-            media_type=None,
-            background=None,
-        )
+        sub_response = Response()
+        if sub_response.status_code == 200:
+            # Set default status code to None
+            sub_response.status_code = None  # typing: ignore
```
Hmm... Maybe just doing the same as I did for FastAPI? :thinking:

patrick91 replied:
@TimPansino have you seen this? 

@Kludex are you talking about this change? https://github.com/tiangolo/fastapi/pull/4483/files#diff-aef3dac481b68359f4edd6974fa3a047cfde595254a4567a560cebc9ccb0673fL465

Kludex replied:
yes

TimPansino replied:
I like it, much simpler.

Kludex commented on strawberry/asgi/handlers/http_handler.py:
```diff
@@ -54,8 +50,6 @@ async def handle(self, scope: Scope, receive: Receive, send: Send):
             context=context,
         )
 
-        response.headers.raw.extend(sub_response.headers.raw)
```
This can be readded, can't it?

TimPansino replied:
Good call, didn't mean to delete that. Fixed.

DoctorJohn commented on strawberry/asgi/handlers/http_handler.py:
```diff
@@ -35,13 +35,9 @@ async def handle(self, scope: Scope, receive: Receive, send: Send):
-            background=None,
-        )
+        sub_response = Response()
+        sub_response.status_code = None  # typing: ignore
```
```suggestion
        sub_response.status_code = None  # type: ignore
```
This should satisfy the mypy checks :)

patrick91 commented:
FastAPI is taking a long time to add support for latest starlette, so I think we should try to support both versions for a while :)

Tushant commented:
latest starlette is now supported in fastapi.

Kludex commented:
> latest starlette is now supported in fastapi.

it's not, latest starlette is 0.20.0... but starlette 0.18.0 is.

krauss commented:
Hey folks, do we have any news on this?

AntonOfTheWoods commented:
`FastAPI` 0.77.1 is out and supports `starlette` 0.19.1. Does that mean this PR needs an update?

patrick91 commented:
@AntonOfTheWoods I'll update this PR soon, I'm a bit busy these days 😊

patrick91 commented on pyproject.toml:
```diff
@@ -27,7 +27,7 @@ build-backend = "poetry.masonry.api"
 [tool.poetry.dependencies]
 python = "^3.7"
-starlette = {version = ">=0.13.6,<0.17.0", optional = true}
+starlette = {version = ">=0.13.6,<0.19.0", optional = true}
```
```suggestion
starlette = {version = ">=0.13.6", optional = true}
```

patrick91 commented:
@Kludex does this PR look good to you? 😊

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Timothy Pansino for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```
<!-- action-check: release-file -->

Kludex commented:
Yep. If you're sure it works with Starlette <0.17.0, then it's fine 👍

AntonOfTheWoods commented:
Yay! :-)

Kludex commented:
Is it an option to have one entry with latest Starlette without FastAPI installed, and skip the tests on FastAPI if not installed?

patrick91 commented:
> Is it an option to have one entry with latest Starlette without FastAPI installed, and skip the tests on FastAPI if not installed?

one entry of tests? 

I tried that in this commit, but it didn't work: https://github.com/strawberry-graphql/strawberry/pull/1594/commits/8d553f79f72b89c4e1db3a307c73217414822ccb

I'm not sure if there's a way to force poetry to install a package, maybe I can try with pip directly, let me see.

patrick91 commented:
> > Is it an option to have one entry with latest Starlette without FastAPI installed, and skip the tests on FastAPI if not installed?

> I'm not sure if there's a way to force poetry to install a package, maybe I can try with pip directly, let me see.

well, that sees to have worked! thanks @Kludex

patrick91 commented on RELEASE.md:
```diff
@@ -0,0 +1,7 @@
+Release type: minor
+
+This release adds support for Starlette 0.18 and 0.19
```
```suggestion
This release adds support for Starlette 0.18 to 0.20
```

botberry commented:
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥
